features/332-sso/api.md

```diff
@@ -0,0 +1,311 @@
+
+function getLoginRedirectResult(): Promise<User>;
+function getLinkRedirectResult(): Promise<User>;
+```
+
+### Example
+
+```typescript
+skygear.auth.loginOAuthProviderWithRedirect("google").catch(error => {
+  // If we reach here, something went wrong and window.location.href is unchanged.
+});
+
+// If the above promise did not reject, the page was redirected so the `then` handler
+// will not be called.
+
+// Later on when the flow finishes, the user will be redirected to `callbackURL`.
+
+// Call this to retrieve the result of the flow.
+skygear.auth.getLoginRedirectResult().then(user => {
+  // login suceedeed
+});
+```
+
+# External access token flow
+
+This is not a flow defined by OAuth 2.0. The main usecase is for native applications that have already acquired an access token, for example, via [Facebook Login for iOS](https://developers.facebook.com/docs/facebook-login/ios/).
+
+## Possible Vulnerability
+
+Suppose Alice grants access to Mallory, a malicious application, via OAuth 2.0. Mallory now has the bearer access token of Alice, which is considered a proof of identity. Mallory can now impersonate Alice and login our application with the identity of Alice.
+
+Since this API is vulnerable by design, it is disabled by default. The developer is strongly recommended to use the standard OAuth 2.0 Authorization Code flow even the application is native and has already had acquired an access token.
+
+### API
+
+```typescript
+function loginOAuthProviderWithAccessToken(providerID: string, accessToken: string, options?: LoginOptions): Promise<User>;
+function linkOAuthProviderWithAccessToken(providerID: string, accessToken: string): Promise<User>;
+```
+
+### Example
+
+```typescript
+getAccessTokenViaFacebookSDKSomehow().then(accessToken => {
+  loginOAuthProviderWithAccessToken("facebook", accessToken).then(user => {
+    // login suceedeed
+  });
+});
+```
+
+# Remove OAuth Provider
+
+If the removal will result in no login principals, then removal is forbidden. Such case happens when the user logged in with Google and has no other login means.
+
+### API
+
+```typescript
+function removeOAuthProvider(providerID: string): Promise<void>;
+```
+
+### Example
+
+```typescript
+askConfirmation().then(confirmed => {
+  if (confirmed) {
+    skygear.auth.removeOAuthProvider("google");
+  }
+});
+```
+
+# Merging existing user for login
+
+Some users may forget that they have signed up with email long ago. When they return to the application again, they may login with OAuth 2.0, creating a new user identity unexpectedly. To mitigate this issue, the developer can instruct the Auth gear on how to merge existing user.
+
+All login APIs accept `LoginOptions`. By default, merging is enabled. If `mergeRealm` is not in `allowedRealms`, it is an error and the login process is aborted. Otherwise, the implementation will merge existing user by inferring email from the provider user info and use it and the realm to match existing user.
```

Comment:
I think this option should exist as a server side config; app developer may want to disable auto-linking.
Also, I think this option should default to `false`; it may have security implications. Consider this case:
1. A signup a legit website using B's email.
2. A send a link to the website to B and ask B to sign up/sign in.
3. B sign up with SSO and merged with user created by A.
4. A can now access B's SSO profile.

Reply:
I should have stated why this is a client side option. If the application is configured `allowedRealms` to be `["teacher", "student"]`. The client should have a UI to let the user to select the realm and a button to login with a provider. So this option is part of the user input and cannot be configured at server side.

For the security implication, @chpapa any thoughts?

Reply:
I should clarify the option I refer to is `mergeExistingUser`; I don't oppose allowing `mergeRealm` at client side.

Reply:
Agreed for `mergeRealm`.

For `mergeExistingUser`, there was a discussion before on the use cases, forgive me for the laziness to dig out the related issues, but the conclusion was basically, we need to support the following use cases:

1. Default: no merging of users, non-unique email addresses
    1. Have an API for client-side to indicate it is an email address logged in, so the application can redirect the user to login to their original account, and link the SSO back to the existing account.
    1. Alternatively, the application can be designed in a way it ignores the above or let users to choose, and have multiple accounts even with identical email addresses for different SSO.
1. No merging of users, unique email address
    1. Same with above, except we would stop people 
1. Always auto-merge users login if email addresses match. (It will have security implication, but some applications think this trade-off worth they can still do it)

So basically the idea is to have a server side config, but need some changes of the API to support use case 1.i and 2, would be great if we can consider it as well.

To clarify, basically use case 1.i is like the following, from a user's perspective:

1. User A already have an account, say email address `a@a.com`
2. User A login with another SSO provider, say gmail, with `a@a.com`
3. User A see a screen say it already have an account, he can choose to login first
4. User A login with his account as usual
5. User A link the SSO provider after login.

Reply:
> mergeExistingUser

Both the server and the client has `mergeExistingUser`. If it is true in server config, then merging is enabled. The client can only turn it to false to disable merging per request. If it is false in server config, then the client must set it to false only, setting to true will result in error. `mergeExistingUser` is false by default in server.

> support use case 1.i and 2

We can include the email in the error info so the developer can redirect the user to the login screen with email prefilled. Though I am not sure if it is appropriate to reveal the email in this way.

Or we just return an error complaining user duplicated without any email. Then it is the developer's responsibility to catch this error and act accordingly. I prefer this option.

Reply:
Sure @louischan-oursky , can you write that flow into spec so we make sure our APIs actually support this flow nicely?

Reply:
Sure.